### DISCUSSION TITLE
In app messaging/broadcast framework and approach

### DISCUSSION THREAD

**Discussion by maimoonak** [opensrp/fhircore#455] [Ideas]
In Android we have different ways to communicate (across component messaging) within different components. Since Android recommends async operations so it becomes more important to have such framework for apps using the library

1- Broadcasts - The services on accomplishing certain tasks send a broadcast which consumer-apps can subscribe to, and get notified and act according to their business needs
2- CallBacks - The consumer-apps register a callback before calling the service and get notified when task is done
3- Events - Event based programming is another approach where consumer-app listens to specific events of interest and do their business accordingly

As a consumer-app fhircore app would need to get notified about many events including SyncStarted, SyncSuccess, SyncFailed, DataInserted, DataUpdated, FhirEngineInitilized, FormSaved, FormFailed, FormCancelled, etc etc

I do not see any such messaging framework implemented in fhirsdk (Or may be could not figureout any). Which one is the best approach and what we are planning to implement. Basically rightnow I need this to implement one of the feature where we want to keep user posted about last sync performed however, down the road with new features we may need to have a full fledged messaging flow

**Comment by pld** [comment]
If there's a related approach on the Android FHIR SDK let's make sure to align w/that

**Comment by fredhersch** [comment]
Sync use cases:
1. Show the last sync date in the UI. 
2. Show an icon if there are any issues i.e if the last Sync Failed
3. Show any error messages related to Sync

Discussed on the call:
* Makes sense to add lastUpdated method to the sync API
* Set up a separate call to discuss adding support for this into the SDK. Need to flesh out details further
* @maimoonak keen to work on this within fhirsdk

cc: @jingtang

**Comment by maimoonak** [comment]
https://developer.android.com/topic/libraries/architecture/workmanager/how-to/intermediate-progress

In previous call we discussed the use of LiveData. The CoroutineWorker basically allows to observe the progress as above. The PeriodicSyncWorker would need to be updated to accommodate it.

**Comment by pld** [comment]
This all sounds like work to be done in Android FHIR SDK, in FHIR Core we'd use the API the SDK exposes

**Comment by pld** [comment]
Unless I'm missing something? Maybe we can split this issue and keep this one to include the using the API pieces that stay here in FHIRCore, and a new (or reference to an existing one if it's there) issue in Android FHIR SDK for the other changes

**Comment by f-odhiambo** [comment]
What's the status of this issue?  @maimoonak

**Comment by maimoonak** [comment]
This one, i.e. #142, #175 , [#554](https://github.com/google/android-fhir/issues/554) all are related. The details on the progress can be tracked on [#554](https://github.com/google/android-fhir/issues/554) and the PR [#648](https://github.com/google/android-fhir/pull/648). There were back and forth discussions with Jing and now need a review and feedback on this PR
